addons/udes_hierarchical_tree_view/static/src/hierarchical_tree_view/hierarchical_tree_controller.js

```diff
@@ -0,0 +1,89 @@
+odoo.define("udes_hierarchical_tree_view.HierarchicalTreeController", function (require) {
+  "use strict";
+
+  var AbstractController = require("web.AbstractController");
+
+  var HierarchicalTreeController = AbstractController.extend({
+    custom_events: _.extend({}, AbstractController.prototype.custom_events, {
+      tree_item_clicked: "_onTreeItemClicked",
+      change_item_tree: "_onChangeItemTree",
+      open_form_view: "_onOpenFormView",
+    }),
+
+    /**
+     * @override
+     * @param parent
+     * @param model
+     * @param renderer
+     * @param {Object} params
+     */
+    init: function (parent, model, renderer, params) {
+      this._super.apply(this, arguments);
+    },
+
+    /**
+     * @param {Object} ev
+     * @param {Object} ev.data contains the payload
+     */
+    _onTreeItemClicked: async function (ev) {
+      ev.stopPropagation();
+      if (ev.data.data.children === undefined) {
+        await this.model.expandChildrenOf(
+          ev.data.data.id,
+          ev.data.data.parent_path
+        );
+      } else {
+        this.model.toggleChildrenVisibleForItem(ev.data.data);
+      }
+      this.update({}, { reload: false });
+    },
+
+    /**
+     * @param {Object} ev
+     * @param {Object} ev.data contains the payload
+     */
+    _onOpenFormView: async function (ev) {
+      ev.stopPropagation();
+      if (ev.data.data.id !== undefined) {
+        let form_views = this.actionViews.filter(function (view) { return view.type == "form" });
+        if (form_views.length == 1) {
+          await this.model.do_action({
+            "res_id": ev.data.data.id,
+            "res_model": this.model.modelName,
+            "type": "ir.actions.act_window",
+            "views": [[form_views[0].viewID, "form"]],
+            'context': this.model.loadParams.context,
+          })
+        } else {
+          // TODO(palabaster): Log warning/alertbox, no form view defined.
```

Comment:
Q: Does this mean a form view has to be defined in the action? (i.e. can't fall back to the default form view, the one with the lowest sequence)

Not an issue at all for this story, but might be worth raising a low priority issue for future. :)

Reply:
We could always change that to make an RPC to get the default form on view initialization, but i didn't know off the top of my head how that is done

Reply:
i've added the error now, only developers should ever see it, if we add it to a new view and we don't do the above in xml they will get a warn in the ui, but adding to the xml should always result in one being found. the RPC thing could be a future improvement though so we don't have to do it